### PR TITLE
Delete review app workflow template

### PR DESCRIPTION
**Pull request by neillturner**
## Context
to cleanup when a review app pr is closed 

## Changes proposed in this pull request
- destroying the review app 
- delete all versions of the terraform state in the azure storage account 

see also https://github.com/DFE-Digital/github-actions/pull/105

## Guidance to review

review code and 
see tested on a branch in https://github.com/DFE-Digital/refer-serious-misconduct/tree/delete-review-app
by running 
gh workflow run  delete-review-app.yml -r "delete-review-app" -f pr_number=1228
and you can see the action log at 
https://github.com/DFE-Digital/refer-serious-misconduct/actions/runs/11820195441

also test by running: 

make new_service SERVICE_NAME='refer-serious-misconduct' SERVICE_SHORT=rsm SERVICE_PRETTY="Refer Serious Misconduct" DOCKER_REPOSITORY='ghcr.io/dfe-digital/refer-serious-misconduct' NAMESPACE_PREFIX=tra DNS_ZONE_NAME='refer-serious-misconduct.education.gov.uk/'

Trello Card [delete review app workflow template](https://trello.com/c/IqDW5aC8/2092-delete-review-app-workflow-template)

## Checklist

- [ ] I have performed a self-review of my code, including formatting and typos
- [ ] I have [cleaned the commit history](https://www.annashipman.co.uk/jfdi/good-pull-requests.html)
- [ ] I have added the `Devops` label
- [ ] I have attached the pull request to the trello card

### CODE REVIEW

**Comment by neillturner** [comment]
[delete review app workflow template](https://trello.com/c/IqDW5aC8/2092-delete-review-app-workflow-template)

File: templates/new_service/.github/workflows/delete-review-app.yml [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            Review app refer serious misconduct deployed to <https://#SERVICE_NAME#-${{ env.PR_NUMBER }}.test.teacherservices.cloud> was deleted
+
+      - uses: Azure/login@v2
+        with:
+          creds: ${{ secrets.azure_credentials }}
+
+      - name: Install Storage Blob Extension
+        run: az extension add --name storage-blob-preview
+
+      - run: |
```
Comment (RMcVelia) [inline templates/new_service/.github/workflows/delete-review-app.yml:64]: can we copy the same format for env vars as https://github.com/DFE-Digital/teacher-services-cloud/blob/main/templates/new_service/.github/workflows/backup-db.yml#L51

File: templates/new_service/.github/workflows/delete-review-app.yml [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        uses: marocchino/sticky-pull-request-comment@v2
+        with:
+          header: aks
+          message: |
+            Review app refer serious misconduct deployed to <https://#SERVICE_NAME#-${{ env.PR_NUMBER }}.test.teacherservices.cloud> was deleted
+
+      - uses: Azure/login@v2
+        with:
+          creds: ${{ secrets.azure_credentials }}
+
```
Comment (RMcVelia) [inline templates/new_service/.github/workflows/delete-review-app.yml:60]: we will need gcp login. e.g. https://github.com/DFE-Digital/apply-for-qualified-teacher-status/blob/main/.github/workflows/delete-review-app.yml#L25 but using repo variables for project_id and wip i.e. vars.project_id and vars.gcp_wip. Only run if both vars aren't empty

File: templates/new_service/.github/workflows/delete-review-app.yml [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+          if [ ${{ github.event_name }} == 'workflow_dispatch' ]; then
+            PR_NUMBER=${{ github.event.inputs.pr_number }}
+          else
+            PR_NUMBER=${{ github.event.pull_request.number }}
+          fi
+          echo "PR_NUMBER=$PR_NUMBER" >> $GITHUB_ENV
+
+      - uses: hashicorp/setup-terraform@v3
+        with:
+          terraform_version: 1.6.4
```
[the remark this answers is not in the document]
Comment (neillturner) [reply templates/new_service/.github/workflows/delete-review-app.yml]: changed

File: templates/new_service/.github/workflows/delete-review-app.yml [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        env:
+          PR_NUMBER: ${{ env.PR_NUMBER }}
+
+      - name: Post Pull Request Comment
+        if: ${{ github.event_name == 'pull_request' }}
+        uses: marocchino/sticky-pull-request-comment@v2
+        with:
+          header: aks
+          message: |
+            Review app refer serious misconduct deployed to <https://#SERVICE_NAME#-${{ env.PR_NUMBER }}.test.teacherservices.cloud> was deleted
```
[the remark this answers is not in the document]
Comment (neillturner) [reply templates/new_service/.github/workflows/delete-review-app.yml]: changed

File: templates/new_service/.github/workflows/delete-review-app.yml (hunk 1 above)
Comment (neillturner) [reply templates/new_service/.github/workflows/delete-review-app.yml]: changed

File: templates/new_service/.github/workflows/delete-review-app.yml [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+          terraform_version: 1.6.4
+          terraform_wrapper: false
+
+      - uses: DFE-Digital/github-actions/set-kubelogin-environment@master
+        with:
+          azure-credentials: ${{ secrets.AZURE_CREDENTIALS }}
+
+      - name: Terraform Destroy
+        run: |
+          make ci review terraform-destroy PR_NUMBER=${{ env.PR_NUMBER }}
```
[the remark this answers is not in the document]
Comment (neillturner) [reply templates/new_service/.github/workflows/delete-review-app.yml]: changed

File: templates/new_service/.github/workflows/delete-review-app.yml (hunk 2 above)
Comment (neillturner) [reply templates/new_service/.github/workflows/delete-review-app.yml:65]: why can  i find the vars.project_id and vars.gcp_wip variables ?

Comment (RMcVelia) [reply templates/new_service/.github/workflows/delete-review-app.yml:65]: we would add them to the repo if it uses gcp e.g. https://github.com/DFE-Digital/apply-for-qualified-teacher-status/settings/variables/actions

Comment (neillturner) [reply templates/new_service/.github/workflows/delete-review-app.yml]: added gcp login

**Review by RMcVelia — COMMENTED** [review]
looks good. As per the other templates we should move this into our github-actions repo (called delete-review-app) then this workflow would just call that.

File: templates/new_service/.github/workflows/delete-review-app.yml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        uses: DFE-Digital/github-actions/delete-review-app@master
+        with:
+          azure-credentials: ${{ secrets.AZURE_CREDENTIALS }}
+          container-name: terraform-state
+          gcp-wip: ${{ vars.GCP_WIP }}
+          gcp-project-id: ${{ vars.GCP_PROJECT_ID }}
+          pr-number: ${{ env.PR_NUMBER }}
+          resource-group-name: "${{ env.AZURE_RESOURCE_PREFIX }}-#SERVICE_SHORT#-${{ env.CONFIG_SHORT }}-rg"
+          storage-account-name: "${{ env.AZURE_RESOURCE_PREFIX }}#SERVICE_SHORT#${{ env.CONFIG_SHORT }}tfsa"
+          tf-state-file: "${{ env.PR_NUMBER }}_kubernetes.tfstate"
```
[the remark this answers is not in the document]
Comment (neillturner) [reply templates/new_service/.github/workflows/delete-review-app.yml:53]: yes in the makefile 
https://github.com/DFE-Digital/teacher-services-cloud/blob/main/templates/new_service/Makefile
we now have 
-backend-config=key=${ENVIRONMENT}_kubernetes.tfstate
but if you look in all the source accounts lots of different names -)

[the remark this answers is not in the document]
Comment (RMcVelia) [reply templates/new_service/.github/workflows/delete-review-app.yml:53]: yes, and for review `$(eval ENVIRONMENT=pr-${PR_NUMBER})` so we should match that with the template

[the remark this answers is not in the document]
Comment (neillturner) [reply templates/new_service/.github/workflows/delete-review-app.yml]: corrected

**Review by RMcVelia — APPROVED** [review]

**Pull request merged**